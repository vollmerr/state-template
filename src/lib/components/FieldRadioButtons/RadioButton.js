import React from 'react';
import T from 'prop-types';
import classNames from 'classnames';

import * as propUtils from '../../utils/propTypes';

// individual radio button, applies state-template styling
const RadioButton = (props) => {
  const {
    name,
    value,
    option,
    inline,
    variant,
    disabled,
    onChange,
    className,
    'aria-invalid': ariaInvalid,
    'aria-describedby': ariaDescribedby,
  } = props;

  const id = `${name}-${option.value}`;
  const inputProps = {
    id,
    name,
    disabled,
    type: 'radio',
    checked: value === option.value,
    onChange: () => onChange(option.value),
    'aria-invalid': ariaInvalid,
    'aria-describedby': ariaDescribedby,
  };

  let cn = classNames([
    'check',
    className,
  ]);

  if (variant) {
    return (
      <label data-test={'field--radio-button'} className={cn} htmlFor={id}>
        <input className={'hidden-up pos-abs top-0 left-0'} {...inputProps} />
        <span className={`btn btn-lg btn-block bg-${variant}--checked rounded-0`}>
          <div>{option.label}</div>
        </span>
      </label>
    );
  }

  cn = classNames([
    'p-l-md m-l-0 m-r-md',
    className,
    inline ? 'form-check-inline' : 'form-check',
  ]);

  return (
    <label data-test={'field--radio-button'} className={cn} htmlFor={id}>
      <input id={id} className={'hidden-up pos-abs'} {...inputProps} />
      <div className={'check-icon-radio'} aria-hidden>
        <i />
      </div>
      <div>{option.label}</div>
    </label>
  );
};

RadioButton.propTypes = {
  /** Name of field */
  name: T.string.isRequired,

  /** Value of radio button */
  value: T.string.isRequired,

  /** Option to select */
  option: propUtils.option.isRequired,

  /** Display inline */
  inline: T.bool,

  /** Use style variant */
  variant: T.oneOf([
    '',
    'highlight',
  ]),

  /** Disable the input */
  disabled: T.bool,

  /** Called when radio button changes */
  onChange: T.func.isRequired,

  /** Class name to attach to top level label */
  className: T.string,

  /** Accessible indicator for errors existing */
  'aria-invalid': T.string.isRequired,

  /** Accessible indicator of related information */
  'aria-describedby': T.string,
};

RadioButton.defaultProps = {
  inline: false,
  variant: '',
  disabled: false,
  className: '',
  'aria-describedby': null,
};

export default RadioButton;
