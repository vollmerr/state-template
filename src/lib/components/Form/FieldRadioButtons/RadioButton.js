import React from 'react';
import T from 'prop-types';
import classNames from 'classnames';

// individual radio button, applies state-template styling
const RadioButton = ({
  input,
  value,
  label,
  variant,
  disabled,
  className,
  inline,
}) => {
  const checked = input.value.toString().indexOf(value) !== -1;
  const onChange = () => input.onChange(value);
  const inputProps = {
    disabled,
    checked,
    onChange,
    type: 'radio',
    name: input.name,
  };

  let cn = classNames([
    'check',
    className,
  ]);

  if (variant) {
    return (
      <label className={cn}>
        <input className={'hidden-up pos-abs top-0 left-0'} {...inputProps} />
        <span className={`btn btn-lg btn-block bg-${variant}--checked rounded-0`}>
          <div>{label}</div>
        </span>
      </label>
    );
  }

  cn = classNames([
    'p-l-md m-l-0 m-r-md',
    className,
    inline ? 'form-check-inline' : 'form-check',
  ]);

  return (
    <label className={cn}>
      <input className={'hidden-up pos-abs'} {...inputProps} />
      <div className={'check-icon-radio'}>
        <i />
      </div>
      <div>{label}</div>
    </label>
  );
};

RadioButton.propTypes = {
  /** Use style variant */
  variant: T.oneOf([
    '',
    'highlight',
  ]),
  /** Input from redux-form's Field, attaches name, value, etc */
  input: T.object.isRequired,
  /** Label of radio button */
  label: T.string.isRequired,
  /** Value of radio button */
  value: T.string.isRequired,
  /** Disable the input */
  disabled: T.bool,
  /** Class name to attach to top level label */
  className: T.string,
  /** Display inline */
  inline: T.bool,
};

RadioButton.defaultProps = {
  variant: '',
  disabled: false,
  className: '',
  inline: false,
};

export default RadioButton;
