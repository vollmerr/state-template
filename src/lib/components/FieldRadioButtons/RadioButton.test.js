import React from 'react';
import { shallow } from 'enzyme';

import RadioButton from './RadioButton';

const props = {
  'aria-describedby': 'aria-desc-id',
  'aria-invalid': 'false',
  disabled: false,
  inline: false,
  name: 'test-name',
  onBlur: jest.fn(),
  onChange: jest.fn(),
  onFocus: jest.fn(),
  option: {
    label: 'label-1',
    value: 'value-1',
    className: 'option-class',
  },
  value: 'value-1',
  variant: null,
};

let wrapper;
describe('RadioButton', () => {
  beforeEach(() => {
    wrapper = shallow(<RadioButton {...props} />);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render as a button if passed a variant', () => {
    wrapper.setProps({ variant: 'highlight' });
    expect(wrapper.find('.btn').length).toBe(1);
  });

  it('should render inline', () => {
    wrapper.setProps({ inline: true });
    expect(wrapper.find('.form-check-inline').length).toBe(1);
  });

  it('should pass the option`s value to the `onChange`', () => {
    wrapper.find('input').simulate('change');
    expect(props.onChange).toBeCalledWith(props.option.value);
  });
});
