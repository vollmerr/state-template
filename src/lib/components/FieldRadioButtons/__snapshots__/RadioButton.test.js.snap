// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButton should render correctly 1`] = `
<label
  className="p-l-md m-l-0 m-r-md form-check"
  htmlFor="testName-v1"
>
  <input
    checked={false}
    className="hidden-up pos-abs"
    disabled={false}
    id="testName-v1"
    name="testName"
    onChange={[Function]}
    type="radio"
  />
  <div
    className="check-icon-radio"
  >
    <i />
  </div>
  <div>
    test label
  </div>
</label>
`;
